src/backend/distributed/utils/metadata_cache.c

```diff
@@ -843,8 +843,17 @@ BuildDistTableCacheEntry(DistTableCacheEntry *cacheEntry)
 	/* note that for reference tables partitionKeyisNull is true */
 	if (!partitionKeyIsNull)
 	{
+		Var *partitionColumn = NULL;
+
 		oldContext = MemoryContextSwitchTo(CacheMemoryContext);
+
+		/* get the string representation of the partition column Var */
 		cacheEntry->partitionKeyString = TextDatumGetCString(partitionKeyDatum);
+
+		/* convert the string to a generic Var */
+		partitionColumn = (Var *) stringToNode(cacheEntry->partitionKeyString);
```

Comment:
I think it's best to have the `Assert` when building the entry and remove all the `Assert`s during access.